src/Compilers/Core/Portable/CommandLine/AnalyzerConfigSet.cs

```diff
@@ -651,10 +651,33 @@ internal sealed class GlobalAnalyzerConfig
 
             internal ImmutableArray<AnalyzerConfig.Section> NamedSections { get; }
 
+            /// <summary>
+            /// The names of <see cref="NamedSections"/>, but having been unescaped via <see cref="TryUnescapeSectionName(string, out string?)"/>.
+            /// This is just a simple cache to avoid repeatedly unescaping the same strings over and over again. A null here means the name
+            /// could not be unescaped.
+            /// </summary>
+            internal ImmutableArray<string?> UnescapedSectionNames { get; }
+
             public GlobalAnalyzerConfig(AnalyzerConfig.Section globalSection, ImmutableArray<AnalyzerConfig.Section> namedSections)
             {
                 GlobalSection = globalSection;
                 NamedSections = namedSections;
+
+                var unescapedSectionNames = ArrayBuilder<string?>.GetInstance();
```

Comment:
Do we actually use the unescaped section name anywhere? Consider just updating the `namedSections` in place with the unescaped name.

Reply:
Hmm thinking about it, maybe we can just do the escaping as we build the global config. I don't think the unescaped name is ever used anywhere, so we can probably just do it once upfront instead.

Reply:
@chsienki I was somewhat worried whether that might introduce new bugs since it means some section names are escaped and some aren't and it's more or less up to you to know what's what. Now that said, there's other code here:

https://github.com/dotnet/roslyn/blob/cc888b28d4b7e5e2050883bc2e44056d6f00c26b/src/Compilers/Core/Portable/CommandLine/AnalyzerConfigSet.cs#L572-L578

That's using the escaped string when trying to merge sections. Looking at our lexer \ is used to escape and character, even if it otherwise doesn't require escaping, so it would mean Foo and F\oo will convert to the same string, but we won't merge the properties because we're dealing with escaped and not unescaped strings.

I can't tell if that's a bug in our lexer that we even allow such a construct (the .editorconfig specification says "Special characters can be escaped with a backslash so they won't be interpreted as wildcard patterns.", which isn't "all characters can be escaped..."), but in general this is arguing for doing the escaping earlier, I think.

Reply:
@jasonmalinowski Yeah, the specification is annoyingly vague on that front. I'm not sure where I land on which way we should interpret it, but think the current behavior is probably fine. If you want a backslash you can escape it, so its at least consistent and keeps the lexer simpler.

That being said global configs are already _not_ editor configs, so I think as long as we decide on the explicit set of rules there (and document them) we're probably ok.

Reply:
@chsienki Yeah, I'm OK keeping the lexer behavior. Does that then mean we have a bug for the `Foo` and `Fo\o` example? Because if we're calling that a bug then that absolutely motivates me to redo this PR to do the unescape process earlier and just update the section name directly since the more look the more I'm finding problems here otherwise.

Reply:
Yeah, I think doing it earlier is the right approach here.

Reply:
FWIW, I just went to check about path separators, and in calling out that `\` is not a valid path seperator it actually lists its as `\\` 
<img width="611" alt="image" src="https://user-images.githubusercontent.com/16246502/153097483-2ce5c496-f2a3-4f0c-8652-951f545acc70.png">

So I think the intention behind the spec is also that a backslash char is *always* an escape character. And we can just process up front and be correct / compliant.

Reply:
Alright, time to rewrite this PR then!

Reply:
Ah, my bad we don't have the bug. Once we're actually looping through global configs we don't break on the first one but keep searching regardless. Still going to simplify this though that way.